From 465abb5562954a8d6431e0128cf7e3d206bf2bcb Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Wed, 12 Aug 2015 15:23:26 +0300
Subject: [PATCH 178/214] dpaa2-eth: Iterate only through initialized channels

The number of DPIO objects available to a DPNI may be fewer than the
number of online cores. A typical example would be a DPNI with a
distribution size smaller than 8. Since we only initialize as many
channels (DPCONs) as there are DPIOs, iterating through all online cpus
would produce a nasty oops when retrieving ethtool stats.

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
[Original patch extracted from FSL LS2085 SDK EAR5.0,
Layerscape2-SDK-SOURCE-20150828-yocto.iso]
Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
index f8495b0..cb79f98 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
@@ -221,7 +221,8 @@ static void ldpaa_get_ethtool_stats(struct net_device *net_dev,
 	}
 	i += j;
 
-	for_each_online_cpu(j) {
+	/* We may be using fewer DPIOs than actual CPUs */
+	for_each_cpu(j, &priv->dpio_cpumask) {
 		ch_stats = &priv->channel[j]->stats;
 		cdan += ch_stats->cdan;
 		portal_busy += ch_stats->dequeue_portal_busy;
-- 
2.5.3

